### ISSUE TITLE
Ability to show/hide and enable/disable spec fields in the UI based on values of other spec fields

### ISSUE THREAD

**Issue by grishick** [airbytehq/airbyte#15809]
## Tell us about the problem you're trying to solve
We need to be able to show/hide or enable/disable certain spec fields based on values of other spec fields. So far we have been working around this limitation with `oneOf` type field, but this workaround does not always work. 

## Use case 1
We would like to allow users to turn off SSL when SSH Tunneling is enabled in order to avoid double encryption and in order to unblock users, who run source databases in special containerized environments that do not allow SSL connections (see [Allow users to turn off SSL on cloud #15265](https://github.com/airbytehq/airbyte/issues/15265) ). 

A possible workaround is to allow user selecting an invalid configuration (SSL = off and SSH Tunnel = off) and then thrown an error in `check` method. This is poor UX, because it involves showing an error message that can be avoided. Today, users of SaaS services are accustomed to configuration UIs that steer them towards valid configurations and prevent selecting invalid options.

## Use case 2
We would like to allow users to select SSL modes `allow` and `prefer` only when SSH Tunneling is enabled. Currently, there is no way to do that. 

## Use case 3
We would like to merge BigQuery and BigQuery Denormalized destination connectors into one. In order to do that, we would have to introduce a field `Enable normalization` to the spec and hide/disable some of the other fields depending on the value of `Enable normalization`

## Use case 4
For authorization options we are duplicating fields under `oneOf` section. This opens up a potential for invalid configurations and complicates maintenance of configuration database, as well as future changes to connector specs.

## Use case 5
For SSL mode specs we are also duplicating several certificate-related fields under `oneOf` section. 

## Additional context
AFAIK, ReactJS provides all the tools necessary for implementing this UI logic. At the very basic level, we could add a spec property that will evaluate to a boolean value based on other spec values. Here's a theoretical example of a spec field that hides SSL options of SSH Tunneling is disabled:

```
"ssl": {
        "title": "SSL Connection",
        "description": "Encrypt data using SSL.",
        "type": "boolean",
        "default": true,
        "order": 6
         "conditions": {
              "enabled":"${tunnel_method} != 'no_tunnel'" 
         },
      }
```

**Comment by octavia-squidington-iii** [comment]
cc @airbytehq/frontend

**Comment by timroes** [comment]
Thanks for filing this. Before this could be worked on we'd need to properly spec out how this syntax should look alike.

I'm not entirely seeing how Use-Case #2 would be solved by this? Are we looking (as the issue suggests) into show/hiding/enabling/disabling specific fields, or actually specific values inside fields (which to my understanding is what Use-Case #2 describes)?

Some thoughts around the final syntax:

**Language vs structured data**

We should not try to come up with a language in a string (like in the above example). This means we'd need to write a parser in the FE that's able to parse that language, which is a significant amount of work to write and maintain. I think we should try to go as much as possible for structured formats (i.e. more like `{ field_name: string, comparator: 'eq' | 'neq', value }`), since that will allow us to work of that values directly without requiring a language parser.

**JSON Schema**

The `connectionSpecification` is a JSON Schema (version 7) at the moment. I didn't find `conditions` anywhere in the spec? I think we should try to avoid adding custom flags into the JSON Schema, since that will make it harder again to parse/type. We should make sure that really none of the [conditional features](https://json-schema.org/understanding-json-schema/reference/conditionals.html) in JSON schema does fulfill our needs before we add something non-standard into those objects. (We already have `airbyte_type` in there which is non-standard, and this is already a code smell, since we need to overwrite types to access it in some places. We should be really careful about adding more of those custom properties.)

**Comment by grishick** [comment]
> I'm not entirely seeing how Use-Case https://github.com/airbytehq/airbyte/pull/2 would be solved by this? Are we looking (as the issue suggests) into show/hiding/enabling/disabling specific fields, or actually specific values inside fields (which to my understanding is what Use-Case https://github.com/airbytehq/airbyte/pull/2 describes)?

@timroes you are correct. This use case requires ability to show/hide values rather than fields. So, yes, somewhat different use case.

> Also would it be safe to send hidden/disabled fields still, so they are not getting lost if the user later on might want to change a setting back (and would still have previously entered field values around), or are there some limitation, that disabled or hidden fields should not be sent when creating the connector?

Yes, I think it would be safe to send hidden/disabled fields and the connector should ignore them

> JSON Schema

We may be able to achieve **Use case 2** with `if/then/else` if we add an ability to have a hidden property or a disabled control (this is a simplified version, because the whole schema is too long):
```
  "if": {
    "properties": { "tunnel_method": { "const": "NO_TUNNEL" } }
  },
  "then": {
      "properties": { 
          "ssl": {
              "type": "boolean",
              "default": true,
              "disabled": true
          },
          { "ssl_mode": { "oneOf": ["require", "verify", "verify_full"] } }
  },
  "else": {
    "properties": { 
        "ssl": {
            "title": "Connect using SSL",
            "type": "boolean",
            "disabled": false
            "default": false,
            "order": 7
        },
        "ssl_mode": {
            "type": "object",
            "title": "SSL Modes",
            "oneOf": ["disable", "allow", "prefer", "require", "verify", "verify_full"] 
        } 
    }
  }
```

given the above schema, will the UI be able to hide or disable "ssl" and show only 3 options for "ssl_mode" if "tunnel_method" is set to "NO_TUNNEL" and otherwise show "ssl" and show all 6 options for "ssl_mode"?
